drivers/tapdisk-image.c

```diff
@@ -313,14 +363,20 @@ __tapdisk_image_open_chain(int type, const char *name, int flags,
 	list_add_tail(&image->next, &head);
 
 	if (unlikely(prt_devnum >= 0)) {
-		char dev[32];
-		snprintf(dev, sizeof(dev),
-			 "%s%d", BLKTAP2_IO_DEVICE, prt_devnum);
-		err = tapdisk_image_open(DISK_TYPE_AIO, dev,
-					 flags|TD_OPEN_RDONLY, encryption, &image);
+		err = find_parent_nbd_path(prt_devnum, &prt_nbd_path);
+		if (err != 0) {
+			ERR(err, "Failed to find NBD path for parent, %d", prt_devnum);
+			goto fail;
+		}
+		/* Don't open parent NBD in secondary mode*/
+		flags &= ~TD_OPEN_SECONDARY;
+		err = tapdisk_image_open(DISK_TYPE_NBD, prt_nbd_path,
+					 flags | TD_OPEN_RDONLY | TD_USE_NEW_NBD,
+					 encryption, &image);
 		if (err)
 			goto fail;
 
+		free(prt_nbd_path);
```

Comment:
This is very fragile.  It only doesn't double-free because there's no `goto err` between this and leaving the function.

I'd suggest setting to NULL explicitly, and then any future rearranging of the function will remain safe.